sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+        :attr:`PRIVILEGES` can be used as a default value.
+        """
+
+    def get_mode_type(self, mode: str) -> str:
+        """Retrieve the type of ``mode``.
+
+        :raise ModeTypeUnknown: if the mode's type cannont be determined
+        :return: the mode's type as defined by :attr:`chanmodes`
+
+        ::
+
+            >>> mm = ModeMessage({'A': tuple('beI'), 'B': tuple('k')}, {})
+            >>> mm.get_mode_type('b')
+            'A'
+            >>> mm.get_mode_type('k')
+            'B'
+
+        This method will raise a :exc:`ModeTypeUnknown` if the mode is unknown,
+        including the case where ``mode`` is actually a user privilege.
+        (such as ``v``)
+        """
+        for letter, modes in self.chanmodes.items():
+            if mode in modes:
+                return letter
+        raise ModeTypeUnknown(mode)
+
+    def get_mode_info(
+        self,
+        mode: str,
+        is_added: bool
+    ) -> Tuple[Optional[str], bool, bool]:
+        """Retrieve ``mode``'s information when added or removed.
+
+        :raise ModeTypeUnknown: when the mode's type is unknown and isn't a
+                                user privilege
+        :raise ModeTypeImproperlyConfigured: when the mode's type is known but
+                                             there is no information for
+                                             parameters (if and when they are
+                                             required by the mode)
+        :return: a tuple with three values: the mode type, if it requires a
+                 parameter, and if it's a channel mode or a user privilege
+
+        ::
+
+            >>> chanmodes = {'A': tuple('beI'), 'B': tuple('k')}
+            >>> t_params = {'A': PARAM_ALWAYS, 'B': PARAM_ADDED}
+            >>> mm = ModeMessage(chanmodes, t_params)
+            >>> mm.get_mode_info('e', False)
+            ('A', True, False)
+            >>> mm.get_mode_info('k', False)
+            ('B', False, False)
+            >>> mm.get_mode_info('v', True)
+            (None, True, True)
+
+        .. note::
+
+            A user privilege ``mode`` doesn't have a type so the first value
+            returned will be ``None`` in that case.
+        """
+        try:
+            letter = self.get_mode_type(mode)
+        except ModeTypeUnknown:
+            if mode in self.privileges:
+                # a user privilege doesn't have a type
+                return None, True, True
+            # not a user privilege: re-raise error
+            raise
+
+        if letter not in self.type_params:
+            # we don't know how to handle this type of mode
+            raise ModeTypeImproperlyConfigured(mode, letter)
+
+        type_param = self.type_params[letter]
+        return letter, not type_param == PARAM_NEVER and (
+            type_param == PARAM_ALWAYS
+            or type_param == PARAM_ADDED and is_added
+            or type_param == PARAM_REMOVED and not is_added
```

Comment:
This really makes me think that `type_params` should be a bitfield table. It should be much faster to bitwise-`&` the values than to do a string comparison, in theory.

I also just like having excuses to use bitwise logic…

Reply:
That's because `type_param` should be symbols, not int or str value.